Branch name acquiring from .gitreview file with fallback to local branch

Amberize opened the pull request:
During "Push to gerrit" stage branch name is taken from .gitreview file. If file or property does not exist it will fallback to local branch name as it always was before. Such approach could be handful while using feature-branch workflow.

Commit:
Implemented branch name acquiring from .gitreview file with fallback to current branch

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -84,7 +91,25 @@ public void initialized() {
+                properties.load(new FileInputStream(gitReviewFilePath));
+
+                branchName = properties.getProperty("defaultbranch");
+            } catch (IOException e) {
```
I think checking if the file exists would be a bit nicer instead of catching the exception. Also, the catch block could be empty, because you do the branchName assignment anyway when it is empty a few lines below.

Please extract this reading code into a new method.

Amberize replied:
Thank you for reviewing my PR. Implemented all your suggestions.

uwolfer reviewed (changes requested):
Thanks a lot for your PR.

Commit:
Extracted gitreview branch name getting into separate method.
Added file existence checking.
Fixed unclosed stream issue.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -21,25 +21,32 @@
 import com.intellij.uiDesigner.core.GridConstraints;
 import com.intellij.uiDesigner.core.GridLayoutManager;
 import com.intellij.util.ui.UIUtil;
+import org.apache.commons.lang.StringUtils;
```
I prefer to use the Guava Joiner in this project (or when a project depends on a recent JDK of course native functionalities).

Amberize replied:
Thanks for pointing me to Guava Joiner. It provides a bit prettier API for string joining purposes.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -84,7 +91,11 @@ public void initialized() {
-            branchTextField.setText(gerritPushTargetPanels.values().iterator().next());
+            String branchName = gerritPushTargetPanels.values().iterator().next();
+
+            String gitReviewBranchName = getGitReviewBranchName();
```
Try to return a Guava Optional here, instead of null.

Amberize replied:
Implemented optional approach for returning gitreview branch name. I was lack of experience with Guava before, so didn't know that there is a possibility to use optional. If it was java 8 I would certainly prefer using built-in optional.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -95,6 +106,29 @@ public void run() {
+                properties.load(fileInputStream);
+                branchName = properties.getProperty("defaultbranch");
+
+                fileInputStream.close();
```
Move this to the finally block of this try-catch.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -95,6 +106,29 @@ public void run() {
+
+                fileInputStream.close();
+            } catch (IOException e) {
+            }
```
If we really want to ignore this exception (which is an option in this case), please add a comment, or in this case if we return a Optional, just "Optional.absent()".

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -95,6 +106,29 @@ public void run() {
+
+                Properties properties = new Properties();
+                properties.load(fileInputStream);
+                branchName = properties.getProperty("defaultbranch");
```
return Optional.fromNullable(properties.getProperty("defaultbranch"))

uwolfer reviewed (changes requested):
Thanks for your update. I have posted a few more suggestions.

Commit:
Switched to Guava Joiner.
Using Guava Optional for gitreview branch name extraction.
Improved exception handling.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java:
```diff
@@ -95,6 +106,37 @@ public void run() {
+        Optional<String> branchName = Optional.absent();
+
+        String gitReviewFilePath = Joiner.on(File.separator).join(
+            ProjectManager.getInstance().getOpenProjects()[0].getBasePath(), GITREVIEW_FILENAME);
```
Thanks for your updates!

I've just looked a bit deeper into this class and I think there is one think we could improve: instead of just taking the first open project here, we could pass the project to `registerGerritPushTargetPanel` (you can get it from the GitRepository instance where this method is called and keep a reference in the `gerritPushTargetPanels` (probably create a simple class for holding both branchName and project as entry)).

But I'm not really sure anymore how this class works with multiple projects open at the same time. It's not really nice code, but I've not found a better way to do it because there is absolutely no API for extending the push dialog.

Can you please try this and do extensive tests with opening multiple projects at the same time, then close one, try the push dialog again and so on?

Thank you very much for your work! If you do not understand my comment, please just ask.

Amberize replied:
I fixed it by simply taking the project from focused window, which invoked push dialog. If you feel that passing the project to `registerGerritPushTargetPanel` is more common and robust solution, please tell me.

uwolfer replied:
This solution is even better than my suggestion. Great work! Thanks.

Commit:
Project selection switched from ProjectManager to DataManager
as the last provides project from focused window.

uwolfer commented:
This change works fine for me when you open push dialog from commit dialog, but it fails when I open push from VCS menu -> Git -> Push.

This line fails because project data is not available:
https://github.com/uwolfer/gerrit-intellij-plugin/blob/intellij14/src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java#L120

``
java.lang.IllegalArgumentException: Argument for @NotNull parameter 'dataContext' of com/intellij/openapi/actionSystem/DataKey.getData must not be null
        at com.intellij.openapi.actionSystem.DataKey.$$$reportNull$$$0(DataKey.java)
        at com.intellij.openapi.actionSystem.DataKey.getData(DataKey.java)
        at com.urswolfer.intellij.plugin.gerrit.push.GerritPushExtensionPanel.getGitReviewBranchName(GerritPushExtensionPanel.java:120)
        at com.urswolfer.intellij.plugin.gerrit.push.GerritPushExtensionPanel.initialized(GerritPushExtensionPanel.java:103)
        at com.urswolfer.intellij.plugin.gerrit.push.GerritPushOptionsPanel.initPanel(GerritPushOptionsPanel.java:55)
        at git4idea.push.GitPushSupport.createOptionsPanel(GitPushSupport.java)
        at com.intellij.dvcs.push.PushController.createAdditionalPanels(PushController.java:621)
        at com.intellij.dvcs.push.ui.VcsPushDialog.<init>(VcsPushDialog.java:54)
        at com.intellij.dvcs.push.VcsPushAction.actionPerformed(VcsPushAction.java:45)
        at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:258)
        at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:275)
        at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.lambda$actionPerformed$0(ActionMenuItem.java:287)
``

@Amberize: Do you think you can look into it soon? The release which I have pushed yesterday contains this PR already.

Amberize commented:
@uwolfer Sure. I'll start to investigate it today.

Amberize commented:
@uwolfer Didn't manage to reproduce the case you've mentioned, but I suppose that following line

https://github.com/uwolfer/gerrit-intellij-plugin/blob/f53665878959ed0ecde132d6e48057fb6e906a4b/src/main/java/com/urswolfer/intellij/plugin/gerrit/push/GerritPushExtensionPanel.java#L119

could produce null ( as it is resolved as AsyncResult.getResult() ) and then that null is passed into DataKey.getData(...)

So could you please change that line to

DataContext dataContext = DataManager.getInstance().getDataContextFromFocus()**.getResultSync()**;

and retest whether it helps?

uwolfer commented:
@Amberize: Thanks for your investigations. I'm also not able to reproduce this issue in all IntelliJ setups, but when it occurred, it freezed the whole IDE after applying your change. When using a timeout, it did not fix the issue. Please check out my last change (which has been released as hotfix) if it works for you as expected. In my tests, it worked as expected.

Amberize commented:
@uwolfer Your fix handles possible NPE from that line of code, but it started to provide null for DataContext when invoking from hotkey. But I have alrready investigated how to change it in order to aquire the right DataContext both from hotkey and VCS menu. Should I create a new PR with 'related to' mark?

uwolfer commented:
@Amberize: Yes, please create a new PR.
